File: backend/node/sum-service/src/queue.ts
import amqp from "amqplib";

export type PublishConfig = {
	headers?: Record<string, string | number | boolean>;
};

export type ConsumeCallback<T> = (
	message: T,
	metadata: { event: string },
	headers: Record<string, unknown>,
) => Promise<void>;

export type Queue = {
	connect(): Promise<void>;
	setup(topology: unknown): Promise<void>;
	publish(event: string, message: unknown, config?: PublishConfig): Promise<void>;
	consume<T = unknown>(queue: string, callback: ConsumeCallback<T>): Promise<void>;
};

export class RabbitMQAdapter implements Queue {
	connection!: amqp.ChannelModel;
	channel!: amqp.Channel;

	async connect(): Promise<void> {
		this.connection = await amqp.connect("amqp://localhost");
		this.channel = await this.connection.createChannel();
	}

	async setup(topologies: RabbitQMTopology[]): Promise<void> {
		for (const topology of topologies) {
			await this.channel.assertExchange(
				topology.exchange.name,
				topology.exchange.type,
				topology.exchange.config,
			);

			for (const queue of topology.queues) {
				await this.channel.assertQueue(queue.name, queue.config);

				for (const binding of queue.bindings) {
					await this.channel.bindQueue(
						queue.name,
						topology.exchange.name,
						binding,
					);
				}
			}
		}
	}

	async publish(
		event: string,
		message: unknown,
		config?: PublishConfig,
	): Promise<void> {
		this.channel.publish(
			"sum.randomize",
			event,
			Buffer.from(JSON.stringify(message)),
			config,
		);
	}

	async consume<T = unknown>(
		queue: string,
		callback: ConsumeCallback<T>,
	): Promise<void> {
		await this.channel.consume(queue, async (msg) => {
			if (!msg) return;

			const metadata = { event: msg.fields.routingKey };
			const headers = msg.properties.headers ?? {};

			try {
				const input = JSON.parse(msg.content.toString()) as T;
				await callback(input, metadata, headers);
				this.channel.ack(msg);
			} catch (error) {
				console.error(error);
			}
		});
	}
}

type ExchangeType = "direct" | "topic" | "headers" | "fanout";

type Exchange = {
	name: string;
	type: ExchangeType;
	config?: amqp.Options.AssertExchange;
};

type QueueDef = {
	name: string;
	config?: amqp.Options.AssertQueue;
	bindings: string[];
};

export type RabbitQMTopology = {
	exchange: Exchange;
	queues: QueueDef[];
};

File: backend/go/sum-service/queue.go
package main

import (
	"encoding/json"

	amqp "github.com/rabbitmq/amqp091-go"
)

type RabbitMQ struct {
	connection *amqp.Connection
	channel    *amqp.Channel
}

func NewRabbitMQ() *RabbitMQ {
	return &RabbitMQ{}
}

func (r *RabbitMQ) Connect() error {
	conn, err := amqp.Dial("amqp://@localhost")
	if err != nil {
		return err
	}

	r.connection = conn
	r.channel, err = conn.Channel()

	if err != nil {
		return err
	}

	return nil
}

func (r *RabbitMQ) Setup(topologies []Topology) {
	for _, topology := range topologies {
		r.channel.ExchangeDeclare(
			topology.Exchange.Name,
			topology.Exchange.Type,
			true,
			false,
			false,
			false,
			nil,
		)

		for _, queue := range topology.Queues {
			r.channel.QueueDeclare(
				queue.Name,
				true,
				false,
				false,
				false,
				nil,
			)

			for _, binding := range queue.Bindings {
				r.channel.QueueBind(
					queue.Name,
					binding,
					topology.Exchange.Name,
					false,
					nil,
				)
			}
		}
	}
}

type PublishConfig struct {
	Headers *amqp.Table
}

func (r *RabbitMQ) Publish(routingKey string, message any, config PublishConfig) error {
	body, err := json.Marshal(message)

	if err != nil {
		return err
	}

	err = r.channel.Publish(
		"go.sum.randomize", // exchange
		routingKey,         // routing key
		false,              // mandatory
		false,              // immediate
		amqp.Publishing{
			ContentType: "application/json",
			Body:        body,
			Headers:     *config.Headers,
		})

	if err != nil {
		return err
	}

	return nil
}

type ConsumerCallbackMessage = []byte
type ConsumerCallbackHeader = map[string]interface{}
type ConsumerCallbackMetadata = map[string]interface{}

type ConsumerCallback func(ConsumerCallbackMessage, ConsumerCallbackHeader, ConsumerCallbackMetadata) error

func (r *RabbitMQ) Consume(queue string, handle ConsumerCallback) {
	incomingMessages, err := r.channel.Consume(
		queue, //queue
		"",    // consumer-tag
		false, // auto-ack
		false, // exclusive
		false, // no-local
		false, // no-wait
		nil,   // args
	)

	if err != nil {
		panic(err)
	}

	go func() {
		for message := range incomingMessages {
			headers := message.Headers

			metadata := make(map[string]interface{})
			metadata["routing-key"] = message.RoutingKey

			err := handle(message.Body, headers, metadata)

			if err != nil {
				message.Nack(false, true)
			}

			message.Ack(false)
		}
	}()
}

type Exchange struct {
	Name string
	Type string
}

type Queue struct {
	Name     string
	Bindings []string
}

type Topology struct {
	Exchange Exchange
	Queues   []Queue
}
